File: src/main/java/tooko/twitter/actions/FollowTask.java
package tooko.twitter.actions;

import tooko.Launcher;
import tooko.main.Fn;
import tooko.twitter.TwitterAccount;
import twitter4j.*;

import java.util.*;

public class FollowTask extends TimerTask {

    public static Timer timer;

    public static void start() {

        stop();

        timer = new Timer("Twitter Follow Task");

        timer.schedule(new TrackTask(), new Date(System.currentTimeMillis() + 60 * 1000L), 3 * 60 * 1000L);

    }

    public static void stop() {

        if (timer != null) {

            timer.cancel();

        }

        timer = null;

    }

    static void fetchInfo(TwitterAccount account, Twitter api) {

        try {

            ResponseList<Status> timeline = api.getHomeTimeline(new Paging().count(200));

            HashSet<Long> queue = new LinkedHashSet<>();

            for (Status status : timeline) {

                queue.add(status.getUser().getId());

                if (status.getInReplyToUserId() > 0) {

                    queue.add(status.getInReplyToStatusId());

                }

            }

            Iterator<Long> iter = queue.iterator();

            while (iter.hasNext()) {

                long accountId = iter.next();

                if (TrackTask.friends.arrayIsIn(account.accountId, "array", accountId) || TrackTask.followers.arrayIsIn(account.accountId, "array", accountId) || AutoData.DATA.arrayIsIn(account.accountId, "autoFollowedIDs", accountId)) {

                    iter.remove();

                }

            }

            LinkedList<User> users = Fn.fetchUsers(api, queue);

            for (User user : users) {


            }

        } catch (TwitterException e) {

            Launcher.log.warn(e);

        }

    }

    @Override
    public void run() {

        List<TwitterAccount> accounts = TwitterAccount.DATA.getAllByField("follow", true);

        for (TwitterAccount account : accounts) {

            fetchInfo(account, account.mkApi());

        }

    }

}

File: src/main/java/tooko/twitter/follow/FollowTask.kt
package tooko.twitter.follow

import cn.hutool.log.LogFactory
import tooko.Launcher
import tooko.main.Fn
import tooko.main.Lang
import tooko.twitter.TwitterAccount
import tooko.twitter.actions.TrackTask
import tooko.twitter.archives.UserA
import tooko.twitter.spam.PredictProcess
import tooko.twitter.spam.UserR
import twitter4j.Paging
import twitter4j.Twitter
import twitter4j.TwitterException
import java.util.*

class FollowTask : TimerTask() {

    val log = LogFactory.get("Follow")

    override fun run() {

        TwitterAccount.DATA.getAllByField("follow", true).forEach {

            fetchInfo(it, it.mkApi())

        }

    }

    companion object {

        private var timer: Timer? = null

        fun start() {

            stop()

            timer = Timer("Twitter FC Task").apply {

                scheduleAtFixedRate(FollowTask(), Date(/*System.currentTimeMillis() + 1 * Fn.h*/), 1 * Fn.h)

            }

        }

        fun stop() {

            timer?.cancel()

            timer = null

        }

    }

    fun fetchInfo(account: TwitterAccount, api: Twitter) {

        try {

            log.debug("FOLLOW TASK START FOR ${account.archive().name}")

            val queue: HashSet<Long> = LinkedHashSet()

            api.getHomeTimeline(Paging().count(200)).forEach {

                queue.add(it.user.id)

                if (it.inReplyToUserId > 0) {

                    queue.add(it.inReplyToStatusId)

                }

            }

            log.debug("FETCHED ${queue.size} IDS")

            val iter = queue.iterator()

            while (iter.hasNext()) {

                val accountId = iter.next()

                if (TrackTask.friends.arrayIsIn(account.accountId, "array", accountId) || TrackTask.followers.arrayIsIn(account.accountId, "array", accountId) || AutoData.DATA.arrayIsIn(account.accountId, "autoFollowedIDs", accountId)) {

                    iter.remove()

                }

            }

            log.debug("NEW ${queue.size} IDS")

            Fn.fetchUsers(api, queue).forEach {

                val archive = UserA.save(it)

                if (PredictProcess.predict(api, UserR.predictUser(archive))) {

                    log.debug("SKIP SPAM :${archive.name}")

                    return@forEach

                }

                if (archive.friends > 500 && archive.followers < 20) {

                    log.debug("SKIP SJH: ${archive.name}")

                    return@forEach

                }

                try {

                    api.createFriendship(archive.accountId)

                    AutoData.DATA.arrayInsert(account.accountId, "autoFollowed", AutoData.AutoFollowed(archive.accountId, System.currentTimeMillis()))
                    AutoData.DATA.arrayInsert(account.accountId, "autoFollowedIDs", archive.accountId)

                    Launcher.twitter.postHtml(account.owner.toLong(), "Followed {}", archive.parseInfo(Lang.get(account.owner)))

                    log.debug("Followed ${archive.name}, RETURN")

                    return

                } catch (ex: TwitterException) {

                    return@forEach

                }


            }

        } catch (e: TwitterException) {

            log.warn(e)

        }
    }

}